src/rep/mesh.jl

```diff
@@ -0,0 +1,988 @@
+export TriMesh,
+    GBMesh,
+    gbmeshes,
+    load_trimesh,
+    save_trimesh,
+    get_verts_list,
+    get_verts_packed,
+    get_verts_padded,
+    get_faces_list,
+    get_faces_packed,
+    get_faces_padded,
+    get_edges_packed,
+    get_edges_to_key,
+    get_faces_to_edges_packed,
+    get_laplacian_packed,
+    get_laplacian_sparse,
+    compute_verts_normals_list,
+    compute_verts_normals_packed,
+    compute_verts_normals_padded,
+    compute_faces_normals_list,
+    compute_faces_normals_packed,
+    compute_faces_normals_padded,
+    compute_faces_areas_list,
+    compute_faces_areas_packed,
+    compute_faces_areas_padded
+
+import GeometryBasics, Printf, MeshIO
+import GeometryBasics:
+    Point3f0, GLTriangleFace, NgonFace, convert_simplex, meta, triangle_mesh
+
+import Zygote: @ignore
+
+"""
+    TriMesh
+
+Initialize Triangle Mesh representation.
+
+### Fields:
+
+- `N`                       - Batch size of TriMesh.
+- `V`                       - Maximum vertices per mesh in TriMesh.
+- `F`                       - Maximum faces per mesh in TriMesh.
+- `equalised`               - Bool, indicates all mesh have same verts and faces size.
+- `valid`                   -
+- `offset`                  - Offset indicating number to be added to migrate to 0-indexed system.
+
+- `_verts_len`              - Number of vertices in each mesh of TriMesh.
+- `_verts_list`             - Vertices in list format.
+- `_verts_packed`           - Vertices in packed format.
+- `_verts_padded`           - Vertices in padded format.
+
+- `_faces_len`              - Number of faces in each mesh of TriMesh.
+- `_faces_list`             - Vertices in list format.
+- `_faces_packed`           - Vertices in packed format.
+- `_faces_padded`           - Vertices in padded format.
+
+- `_edges_packed`           - Edges in packed format (according to packed vertices).
+- `_faces_to_edges_packed`  - Faces formed by edges in packed format (according to packed edges).
+- `_laplacian_packed`       - Laplacian sparce matrix in packed format.
+- `_edges_to_key`           - Dict mapping edges tuple to unique key.
+
+### Available Contructor:
+
+- `TriMesh(verts_list, faces_list; offset::Number = -1)`
+- `TriMesh(m::Vector{<:GeometryBasics.Mesh})`
+- `TriMesh(m::GeometryBasics.Mesh)`
+- `TriMesh(m::TriMesh)`
+
+"""
+mutable struct TriMesh{T<:AbstractFloat,R<:Integer} <: AbstractMesh{T,R}
```

Comment:
Reposting my comment from slack chat here

I am still not totally convinced with the storage of the TriMesh type. The Union{Nothing, AbstractArray} is what concerns me. Since we are always recomputing the cache invalidated values makes me wonder why use nothing at all. Always preallocate these values with zeros and store a cache parameter which is of type Bool. As such the type of TriMesh will not be storing abstract values. Ans check using the cache parameter instead of checking for nothings.

Using nothing might have made sense if we were concerned with memory and would like to deallocate things (not sure if julia gc does free this memory), but here we always (re)compute these values and as such we are simply increasing the number of times we have to allocate the same array.

Reply:
From implementation point of view, we will need to add bool for each field and some changes in get_* functions, and rest should work fine.

Reply:
Agreed. And I feel this should improve performance in general given the compiler can't really infer the exact types of the fields rn.

Reply:
I have added Bool in verts and faces.

Reply:
It looks good now.